###  Name: Shriya Bhat
### Reg: 220968020
### Class: DSE A1
### Week 4 q3

In [3]:
import numpy as np
import random
import matplotlib.pyplot as plt

In [4]:
# Function to calculate the number of attacking pairs of queens
def calculate_attacks(board):
    attacks = 0
    n = len(board)
    for i in range(n):
        for j in range(i + 1, n):
            if board[i] == board[j] or abs(board[i] - board[j]) == abs(i - j):
                attacks += 1
    return attacks

# Function to generate a random initial board
def generate_initial_board(n):
    return [random.randint(0, n - 1) for _ in range(n)]

# Simple Hill Climbing Algorithm
def simple_hill_climbing(n):
    current_board = generate_initial_board(n)
    current_attacks = calculate_attacks(current_board)
    iterations = 0

    while True:
        iterations += 1
        neighbors = []
        
        for col in range(n):
            for row in range(n):
                if row != current_board[col]:
                    neighbor = current_board.copy()
                    neighbor[col] = row
                    neighbors.append(neighbor)

        next_board = min(neighbors, key=calculate_attacks)
        next_attacks = calculate_attacks(next_board)

        if next_attacks >= current_attacks:
            break
        
        current_board, current_attacks = next_board, next_attacks

    return current_board, current_attacks, iterations

# Stochastic Hill Climbing Algorithm
def stochastic_hill_climbing(n):
    current_board = generate_initial_board(n)
    current_attacks = calculate_attacks(current_board)
    iterations = 0

    while True:
        iterations += 1
        neighbors = []

        for col in range(n):
            for row in range(n):
                if row != current_board[col]:
                    neighbor = current_board.copy()
                    neighbor[col] = row
                    neighbors.append(neighbor)

        next_board = random.choice(neighbors)
        next_attacks = calculate_attacks(next_board)

        if next_attacks < current_attacks:
            current_board, current_attacks = next_board, next_attacks
        else:
            break

    return current_board, current_attacks, iterations

# Steepest Ascent Hill Climbing Algorithm
def steepest_ascent_hill_climbing(n):
    current_board = generate_initial_board(n)
    current_attacks = calculate_attacks(current_board)
    iterations = 0

    while True:
        iterations += 1
        neighbors = []

        for col in range(n):
            for row in range(n):
                if row != current_board[col]:
                    neighbor = current_board.copy()
                    neighbor[col] = row
                    neighbors.append(neighbor)

        best_neighbor = min(neighbors, key=calculate_attacks)
        best_attacks = calculate_attacks(best_neighbor)

        if best_attacks < current_attacks:
            current_board, current_attacks = best_neighbor, best_attacks
        else:
            break

    return current_board, current_attacks, iterations

# Main execution loop to run experiments for different values of N
def run_experiment(sizes):
    results = {}

    for n in sizes:
        print(f"Running experiments for N={n}\n")
        
        algorithms = {
            "Simple Hill Climbing": simple_hill_climbing,
            "Stochastic Hill Climbing": stochastic_hill_climbing,
            "Steepest Ascent Hill Climbing": steepest_ascent_hill_climbing,
        }

        for name, algorithm in algorithms.items():
            solution, attacks, iterations = algorithm(n)
            results[name] = (solution, attacks, iterations)
            print(f"{name}: \nSolution: {solution}\nAttacking Pairs: {attacks}\nIterations: {iterations}\n")
            
        print()



In [5]:
run_experiment([4,8,16,32,64])

Running experiments for N=4

Simple Hill Climbing: 
Solution: [2, 0, 3, 1]
Attacking Pairs0
Iterations: 2

Stochastic Hill Climbing: 
Solution: [0, 0, 0, 2]
Attacking Pairs4
Iterations: 1

Steepest Ascent Hill Climbing: 
Solution: [1, 2, 0, 3]
Attacking Pairs1
Iterations: 2


Running experiments for N=8

Simple Hill Climbing: 
Solution: [5, 3, 1, 7, 4, 6, 0, 2]
Attacking Pairs0
Iterations: 3

Stochastic Hill Climbing: 
Solution: [5, 7, 2, 0, 6, 6, 5, 3]
Attacking Pairs4
Iterations: 2

Steepest Ascent Hill Climbing: 
Solution: [2, 1, 7, 4, 0, 3, 0, 6]
Attacking Pairs2
Iterations: 3


Running experiments for N=16

Simple Hill Climbing: 
Solution: [4, 12, 15, 9, 14, 6, 1, 7, 2, 14, 11, 8, 10, 3, 13, 0]
Attacking Pairs2
Iterations: 8

Stochastic Hill Climbing: 
Solution: [12, 3, 6, 15, 0, 6, 4, 3, 12, 12, 1, 10, 12, 12, 11, 3]
Attacking Pairs24
Iterations: 2

Steepest Ascent Hill Climbing: 
Solution: [8, 0, 5, 7, 9, 15, 13, 4, 6, 0, 11, 1, 10, 2, 14, 3]
Attacking Pairs2
Iterations: 9


Run

In [12]:
#include execution time too

In [10]:
import random
import time

# Function to calculate the number of attacking pairs of queens
def calculate_attacks(board):
    attacks = 0
    n = len(board)
    for i in range(n):
        for j in range(i + 1, n):
            if board[i] == board[j] or abs(board[i] - board[j]) == abs(i - j):
                attacks += 1
    return attacks

# Function to generate a random initial board
def generate_initial_board(n):
    return [random.randint(0, n - 1) for _ in range(n)]

# Simple Hill Climbing Algorithm
def simple_hill_climbing(n):
    current_board = generate_initial_board(n)
    current_attacks = calculate_attacks(current_board)
    iterations = 0
    restarts = 0

    while True:
        iterations += 1
        neighbors = []
        
        # Generate neighbors by moving each queen to a different row
        for col in range(n):
            for row in range(n):
                if row != current_board[col]:
                    neighbor = current_board.copy()
                    neighbor[col] = row
                    neighbors.append(neighbor)

        next_board = min(neighbors, key=calculate_attacks)
        next_attacks = calculate_attacks(next_board)

        # If no improvement, restart
        if next_attacks >= current_attacks:
            restarts += 1
            return current_board, current_attacks, iterations, restarts
        
        current_board, current_attacks = next_board, next_attacks

    return current_board, current_attacks, iterations, restarts

# Stochastic Hill Climbing Algorithm
def stochastic_hill_climbing(n):
    current_board = generate_initial_board(n)
    current_attacks = calculate_attacks(current_board)
    iterations = 0
    restarts = 0

    while True:
        iterations += 1
        neighbors = []

        # Generate neighbors by moving each queen to a different row
        for col in range(n):
            for row in range(n):
                if row != current_board[col]:
                    neighbor = current_board.copy()
                    neighbor[col] = row
                    neighbors.append(neighbor)

        next_board = random.choice(neighbors)
        next_attacks = calculate_attacks(next_board)

        # If no improvement, restart
        if next_attacks < current_attacks:
            current_board, current_attacks = next_board, next_attacks
        else:
            restarts += 1
            return current_board, current_attacks, iterations, restarts

    return current_board, current_attacks, iterations, restarts

# Steepest Ascent Hill Climbing Algorithm
def steepest_ascent_hill_climbing(n):
    current_board = generate_initial_board(n)
    current_attacks = calculate_attacks(current_board)
    iterations = 0
    restarts = 0

    while True:
        iterations += 1
        neighbors = []

        # Generate neighbors by moving each queen to a different row
        for col in range(n):
            for row in range(n):
                if row != current_board[col]:
                    neighbor = current_board.copy()
                    neighbor[col] = row
                    neighbors.append(neighbor)

        best_neighbor = min(neighbors, key=calculate_attacks)
        best_attacks = calculate_attacks(best_neighbor)

        # If no improvement, restart
        if best_attacks < current_attacks:
            current_board, current_attacks = best_neighbor, best_attacks
        else:
            restarts += 1
            return current_board, current_attacks, iterations, restarts

    return current_board, current_attacks, iterations, restarts

# Main execution loop to run experiments for different values of N
def run_experiment(sizes):
    results = []

    for n in sizes:
        print(f"Running experiments for N={n}\n")
        
        algorithms = {
            "Simple Hill Climbing": simple_hill_climbing,
            "Stochastic Hill Climbing": stochastic_hill_climbing,
            "Steepest Ascent Hill Climbing": steepest_ascent_hill_climbing,
        }

        for name, algorithm in algorithms.items():
            start_time = time.time()  # Start measuring time
            solution, attacks, iterations, restarts = algorithm(n)
            execution_time = time.time() - start_time  # End measuring time
            success = attacks == 0  # Success is when there are no attacking pairs
            
            results.append({
                "Algorithm": name,
                "N": n,
                "Solution": solution,
                "Attacking Pairs": attacks,
                "Iterations": iterations,
                "Restarts": restarts,
                "Execution Time (s)": execution_time,
                "Success": success,
            })

            # Output results
            print(f"{name}: Solution: {solution}")
            print(f"Attacking Pairs: {attacks}")
            print(f"Iterations: {iterations}")
            print(f"Restarts: {restarts}")
            print(f"Execution Time: {execution_time:.4f} seconds\n")

    # Convert results to DataFrame for easier analysis
    import pandas as pd
    df_results = pd.DataFrame(results)
    print(df_results)

# Running the experiment for different N values
run_experiment([4, 8, 16, 32, 64])


Running experiments for N=4

Simple Hill Climbing: Solution: [0, 2, 3, 1]
Attacking Pairs: 1
Iterations: 3
Restarts: 1
Execution Time: 0.0000 seconds

Stochastic Hill Climbing: Solution: [1, 0, 2, 0]
Attacking Pairs: 2
Iterations: 2
Restarts: 1
Execution Time: 0.0000 seconds

Steepest Ascent Hill Climbing: Solution: [1, 3, 0, 2]
Attacking Pairs: 0
Iterations: 4
Restarts: 1
Execution Time: 0.0000 seconds

Running experiments for N=8

Simple Hill Climbing: Solution: [1, 6, 2, 7, 7, 4, 0, 5]
Attacking Pairs: 1
Iterations: 4
Restarts: 1
Execution Time: 0.0010 seconds

Stochastic Hill Climbing: Solution: [0, 4, 3, 4, 6, 1, 1, 6]
Attacking Pairs: 6
Iterations: 1
Restarts: 1
Execution Time: 0.0000 seconds

Steepest Ascent Hill Climbing: Solution: [6, 2, 2, 5, 7, 0, 3, 6]
Attacking Pairs: 2
Iterations: 3
Restarts: 1
Execution Time: 0.0000 seconds

Running experiments for N=16

Simple Hill Climbing: Solution: [9, 15, 12, 4, 7, 1, 13, 6, 2, 6, 14, 11, 0, 8, 3, 5]
Attacking Pairs: 1
Iterations: 8

# Performance Analysis of Hill Climbing Algorithms

## Introduction

The performance of each algorithm is assessed based on the following criteria:

1. **Attacking Pairs**: The number of pairs of queens that attack each other in the solution.
2. **Iterations**: The number of steps taken by the algorithm to reach the solution.
3. **Restarts**: The number of restarts made by the algorithm if a local optima is encountered.
4. **Execution Time**: The time taken by the algorithm to find a solution.
5. **Success**: Whether the algorithm found a solution (True) or not (False).

---

## Observations

### Success Rates
- **Simple Hill Climbing**:
  - For smaller values of N (e.g., N=4, N=8), the Simple Hill Climbing algorithm often finds a solution (with a few attacking pairs).
  - For larger values (N=32, N=64), it fails to find a solution, indicating that it struggles with larger problem sizes.
- **Stochastic Hill Climbing**:
  - This algorithm often produces solutions, but the number of attacking pairs is higher, especially for larger N.
  - The algorithm has a higher tendency to get stuck in suboptimal solutions compared to Steepest Ascent Hill Climbing.
- **Steepest Ascent Hill Climbing**:
  - The Steepest Ascent variant performs the best in terms of minimizing attacking pairs and finding a solution.
  - It consistently produces solutions with zero attacking pairs (N=4, N=8, N=16), which is an indication of its higher effectiveness compared to the other two algorithms.

### Iterations and Restarts
- **Simple Hill Climbing**: 
  - Generally requires more iterations to reach a solution, especially as N increases. It may get stuck in local optima, resulting in longer execution times.
  - Only requires one restart per experiment, showing that it doesn't get stuck often, but still struggles with larger N.
- **Stochastic Hill Climbing**:
  - Stochastic Hill Climbing often finds a solution in fewer iterations but at the cost of more attacking pairs, particularly for larger N (e.g., N=32, N=64).
  - Like Simple Hill Climbing, it also requires only one restart per experiment.
- **Steepest Ascent Hill Climbing**:
  - This algorithm generally requires more iterations than Stochastic Hill Climbing but tends to find better solutions with fewer attacking pairs.
  - As the size of N increases, it still performs well with no restarts needed.

### Execution Time
- **Simple Hill Climbing**:
  - The execution time increases significantly as N increases. For N=64, the execution time is the longest (22.7 seconds), indicating that it becomes inefficient with larger problem sizes.
- **Stochastic Hill Climbing**:
  - The execution time remains extremely low, even for larger N. This suggests that the algorithm's random decisions may lead to faster (but less optimal) results.
- **Steepest Ascent Hill Climbing**:
  - The execution time is slightly higher than Stochastic Hill Climbing but significantly lower than Simple Hill Climbing for larger N. This suggests that the more targeted approach of Steepest Ascent is more efficient compared to Simple Hill Climbing.

---

## Detailed Performance Table

| Algorithm                     | N   | Solution                                | Attacking Pairs | Iterations | Restarts | Execution Time (s) | Success |
|-------------------------------|-----|-----------------------------------------|-----------------|------------|----------|--------------------|---------|
| Simple Hill Climbing           | 4   | [0, 2, 3, 1]                           | 1               | 3          | 1        | 0.0000             | False   |
| Stochastic Hill Climbing       | 4   | [1, 0, 2, 0]                           | 2               | 2          | 1        | 0.0000             | False   |
| Steepest Ascent Hill Climbing  | 4   | [1, 3, 0, 2]                           | 0               | 4          | 1        | 0.0000             | True    |
| Simple Hill Climbing           | 8   | [1, 6, 2, 7, 7, 4, 0, 5]               | 1               | 4          | 1        | 0.0010             | False   |
| Stochastic Hill Climbing       | 8   | [0, 4, 3, 4, 6, 1, 1, 6]               | 6               | 1          | 1        | 0.0000             | False   |
| Steepest Ascent Hill Climbing  | 8   | [6, 2, 2, 5, 7, 0, 3, 6]               | 2               | 3          | 1        | 0.0000             | False   |
| Simple Hill Climbing           | 16  | [9, 15, 12, 4, 7, 1, 13, 6, 2, 6, 14, 11, 0, 8, 3, 5] | 1 | 8 | 1 | 0.0508 | False |
| Stochastic Hill Climbing       | 16  | [4, 6, 15, 2, 11, 4, 14, 5, 9, 14, 2, 12, 13, 0, 5, 13] | 14 | 3 | 1 | 0.0010 | False |
| Steepest Ascent Hill Climbing  | 16  | [5, 13, 0, 10, 4, 7, 15, 2, 12, 1, 6, 8, 11, 14, 3, 9] | 0 | 9 | 1 | 0.0354 | True |
| Simple Hill Climbing           | 32  | [10, 29, 1, 11, 0, 30, 8, 12, 25, 18, 5, 27, 19, 28, 4, 22, 7, 9, 16, 2, 26, 3, 23, 17, 24, 3, 6, 15, 21, 14, 14, 20] | 4 | 13 | 1 | 0.6670 | False |
| Stochastic Hill Climbing       | 32  | [12, 27, 4, 31, 15, 7, 16, 16, 6, 12, 11, 13, 30, 14, 28, 3, 0, 25, 9, 22, 25, 5, 22, 2, 1, 20, 15, 10, 21, 3, 4, 0] | 25 | 1 | 1 | 0.0000 | False |
| Steepest Ascent Hill Climbing  | 32  | [5, 10, 2, 0, 16, 28, 12, 3, 11, 30, 25, 18, 22, 17, 31, 23, 7, 10, 4, 6, 1, 20, 24, 21, 7, 26, 8, 15, 27, 14, 19, 9] | 4 | 14 | 1 | 0.7937 | False |
| Simple Hill Climbing           | 64  | [8, 52, 22, 31, 14, 49, 63, 5, 38, 11, 23, 20, 60, 48, 26, 61, 39, 43, 9, 6, 17, 58, 36, 59, 28, 4, 37, 51, 45, 10, 19, 46, 16, 29, 33, 53, 21, 27, 18, 32, 7, 2, 24, 44, 1, 40, 29, 54, 26, 55, 15, 0, 41, 15, 30, 56, 3, 13, 50, 47, 34, 25, 62, 57] | 5 | 28 | 1 | 22.6985 | False |
| Stochastic Hill Climbing       | 64  | [42, 32, 26, 31, 32, 29, 4, 52, 5, 24, 29, 20, 22, 26, 51, 6, 36, 26, 43, 24, 13, 2, 61, 6, 25, 51, 1, 26, 62, 10, 38, 38, 18, 31, 61, 34, 50, 50, 26, 60, 57, 1, 19, 47, 0, 58, 41, 19, 21, 10, 20, 12, 19, 35, 13, 11, 61, 45, 31, 40, 13, 53, 37, 45] | 74 | 1 | 1 | 0.0000 | False |
| Steepest Ascent Hill Climbing  | 64  | [12, 35, 38, 60, 10, 4, 55, 31, 43, 9, 37, 22, 1, 56, 27, 17, 3, 47, 39, 36, 0, 52, 5, 19, 21, 34, 58, 50, 20, 62, 45, 45, 33, 61, 44, 2, 30, 59, 57, 46, 48, 13, 40, 63, 49, 26, 14, 25, 41, 7, 54, 11, 15, 44, 29, 37, 18, 42, 23, 28, 8, 51, 24, 16] | 4 | 27 | 1 | 21.8886 | False |

---



# Thank You